Bugfix: Fix exempt URI method views

singlerider opened the pull request:
This Pull Request is a response to #27.

This corrects some testing logic loopholes that just happened to have passed, before. There are still issues with the tests, but these corrections allow for my use-cases to become unblocked.

The important part of this PR is the addition of a new `if` statement:

```python
    def process_view(self, request, view_func, view_args, view_kwargs):
        
        # for now there is no role assigned yet and no userinfo defined
        request.roles = []
        request.userinfo = []

        # Checks the URIs (paths) that doesn't needs authentication        
        if hasattr(settings, 'KEYCLOAK_EXEMPT_URIS'):
            path = request.path_info.lstrip('/')
            if any(re.match(m, path) for m in settings.KEYCLOAK_EXEMPT_URIS):
                # Checks to see if a request.method explicitly overwrites exemptions in SETTINGS
                if hasattr(view_func.cls, "keycloak_roles") and request.method not in view_func.cls.keycloak_roles:
                    return None
```

Specific line:

```python
if hasattr(view_func.cls, "keycloak_roles") and request.method not in view_func.cls.keycloak_roles:
```

This logic checks to see if a `request.method` explicitly overwrites exemptions in `SETTINGS.KEYCLOAK_EXEMPT_URIS`. Adding one more check to see if the processing of a given view should occur makes it simple to allow for more complex behaviors on a `ModelViewSet` than either fully exempted or not.

singlerider committed:
Fix tests assertEquals calls

When running tests in Python 3.12.0, an error indicating `assertEquals`
is not available.

singlerider committed:
Modify tests for exempted URI with explicit override in view

The tests were previously testing against the following configuration:

```python
{'GET': ['director', 'judge', 'employee'], 'POST': ['director', 'judge'], 'UPDATE': ['director', 'judge'], 'DELETE': ['director', 'judge'], 'PATCH': ['director', 'judge', 'employee']}
```

It was not possible to override the test cases. The `'POST'` method was
removed from keycloak_roles in order to be able to appropriately test
the intended behavior, as described in the test case, `test_when_some_EXEMPT_URI_is_permitted_on_authentication_with_keycloak_roles_on_view`.

singlerider committed:
Add explicit override logic for exempted URIs in middleware

This will close and fix #27.

Using a simple `if` statement, adding one more check to see if the
processing of a given view should occur makes it simple to allow for
more complex behaviors on a `ModelViewSet` than either fully exempted or
exempted, yet able to be overriden in a View.

singlerider commented:
The tests have all passed after installing the library locally as instructed in the README:

```
~/git/django-keycloak-auth on  bugfix/fix-exempt-uri-method-views via django-keycloak-auth via 🐍 django-keycloak-auth 
➜ python manage.py test django_keycloak_auth.tests
Found 21 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
.....Error obtaining introspect token from endpoint: TEST/realms/TEST/protocol/openid-connect/token/introspect, data {'token': '', 'client_id': 'client-backend', 'client_secret': None},  headers {'content-type': 'application/x-www-form-urlencoded', 'authorization': 'Bearer '}, response error 
Error obtaining introspect token from endpoint: TEST/realms/TEST/protocol/openid-connect/token/introspect, data {'token': '', 'client_id': 'client-backend', 'client_secret': None},  headers {'content-type': 'application/x-www-form-urlencoded', 'authorization': 'Bearer '}, response error 
.Error obtaining list of endpoints from endpoint: TEST/realms/TEST/.well-known/openid-configuration, response error 
Error obtaining list of endpoints from endpoint: TEST/realms/TEST/.well-known/openid-configuration, response error 
...............
----------------------------------------------------------------------
Ran 21 tests in 0.010s

OK
Destroying test database for alias 'default'...
```

marcelo225 commented:
@singlerider, thanks for contributing it.
I gonna check your PR soon!

Commit:
Update version number

singlerider commented:
I wanted to note that I found the deprecation announcement for `assertEquals` from Python `3.3` (which was released in 2012):

https://docs.python.org/3.3/library/unittest.html#deprecated-aliases
![Screenshot_20231028-130313](https://github.com/marcelo225/django-keycloak-auth/assets/6310006/6a824577-2052-4767-abcc-0898d363b4ba)

marcelo225 commented:
> I wanted to note that I found the deprecation announcement for `assertEquals` from Python `3.3` (which was released in 2012):

Right! I've noticed in your PR that you changed in `tests.py` file. 
Good observation!

singlerider committed:
Fix AttributeError for is_api_view check

When attempting to access the `/admin` interface in Django, an
`AttributeError` gets `raise`d:

```
Traceback (most recent call last):
  File "/Users/thomasfex/development/github/ItkBackend/env/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/Users/thomasfex/development/github/ItkBackend/env/lib/python3.8/site-packages/django/core/handlers/base.py", line 171, in _get_response
    response = middleware_method(request, callback, callback_args, callback_kwargs)
  File "/Users/thomasfex/development/github/ItkBackend/env/lib/python3.8/site-packages/django_keycloak_auth/middleware.py", line 91, in process_view
    is_api_view = True if str(view_func.cls.__qualname__) == "WrappedAPIView" else False
AttributeError: 'function' object has no attribute 'cls'
[03-11-23 13:38:09] django.server 500 "GET /admin/ HTTP/1.1" 500 95736
```

A `try`/`except` block successfully allows the access of `/admin` pages.

Co-authored-by: Thomas Cross <tom.bz2@gmail.com>

singlerider committed:
Merge branch 'bugfix/fix-exempt-uri-method-views' of github.com:singlerider/django-keycloak-auth into bugfix/fix-exempt-uri-method-views

marcelo225 commented:
@singlerider, may I merge into the main branch or do you have some more considerations?

singlerider commented:
If you are comfortable, @marcelo225 , I am comfortable. Thank you!

marcelo225 commented:
> If you are comfortable, @marcelo225 , I am comfortable. Thank you!

@singlerider, the lib version has been updated to 0.9.8
thanks for contributing to it.
